src/NServiceBus.Core/Recoverability/RecoverabilityExecutor.cs

```diff
@@ -0,0 +1,116 @@
+﻿namespace NServiceBus
+{
+    using System;
+    using System.Threading.Tasks;
+    using Logging;
+    using Transport;
+
+    class RecoverabilityExecutor
+    {
+        public RecoverabilityExecutor(IRecoverabilityPolicy recoverabilityPolicy, IEventAggregator eventAggregator, DelayedRetryExecutor delayedRetryExecutor, MoveToErrorsExecutor moveToErrorsExecutor, bool transactionsOn)
+        {
+            this.recoverabilityPolicy = recoverabilityPolicy;
+            this.eventAggregator = eventAggregator;
+            this.delayedRetryExecutor = delayedRetryExecutor;
+            this.moveToErrorsExecutor = moveToErrorsExecutor;
+            this.transactionsOn = transactionsOn;
+
+            raiseNotifications = recoverabilityPolicy.RaiseNotifications;
+            delayedRetriesEnabled = delayedRetryExecutor != null;
+        }
+
+        public Task<ErrorHandleResult> Invoke(ErrorContext errorContext)
+        {
+            if (transactionsOn == false || errorContext.Exception is MessageDeserializationException)
+            {
+                return MoveToError(errorContext);
+            }
+
+            return PerformRecoverabilityAction(errorContext);
+        }
+
+        Task<ErrorHandleResult> PerformRecoverabilityAction(ErrorContext errorContext)
+        {
+            var recoveryAction = recoverabilityPolicy.Invoke(errorContext);
+
+            if (recoveryAction is ImmediateRetry)
+            {
+                return RaiseImmediateRetryNotifications(errorContext);
+            }
+
+            // When we can't do delayed retries then just fall through to error.
+            if (recoveryAction is DelayedRetry && !delayedRetriesEnabled)
+            {
+                Logger.Warn("Current recoverability policy requested delayed retry but delayed delivery is not supported by this endpoint. Consider enabling the timeout manager or use a transport which natively supports delayed delivery. Moving to the error queue instead.");
+                recoveryAction = RecoverabilityAction.MoveToError();
+            }
+
+            var delayedRetryAction = recoveryAction as DelayedRetry;
```

Comment:
Don't think it is worth changing it. I came to the conclusion that the current code structure is intention revealing. From a performance perspective I don't see why we should change it either since when we do recoverability we are not exactly on a fast path anyway . Feel free to convince me otherwise.